### PR TITLE
Optimizer: optimize transposes in variety of circumstances

### PR DESCRIPTION
**Pull request by anderspapitto**
This is a first pass to get feedback

### CODE REVIEW

**Comment by anderspapitto** [comment]
now removes nop transposes, consecutive transposes that cancel out, and transposes-into-Gemm (which can just be a parameter to Gemm)

File: torch/csrc/jit/passes/peephole.cpp [hunk 1]
```diff
@@ -20,6 +20,36 @@ void PeepholeOptimize(std::shared_ptr<Graph>& graph) {
...
+      }
+    }
+
+    std::vector<int64_t> simpleTransPerm({1,0});
+
+    // Inverse transposes
+    if (n->kind() == kTranspose &&
+        n->input()->kind() == kTranspose) {
+      if (n->is(kperm) == simpleTransPerm &&
+          n->input()->is(kperm) == simpleTransPerm) {
```
Comment (ezyang) [inline torch/csrc/jit/passes/peephole.cpp]: You can generalize this, right? Transposes in ONNX are secretly permutations, and permutations are closed under composition.

**Comment by ezyang** [comment]
Thanks, this looks great! I wrote one minor comment about generalizing transpose/permutation fusion, but everything looks fine.

Before we merge, we're going to need a test. Do you know how to operate the accept test machinery?

File: torch/csrc/jit/passes/peephole.cpp (hunk 1 above)
Comment (anderspapitto) [reply torch/csrc/jit/passes/peephole.cpp]: yeah I was thinking this too. I'll take a stab at it.

**Comment by anderspapitto** [comment]
generalized logic, cleaned up code and comments, and corrected several bugs related to mutating shared references and to resource deallocation.

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 2]
```diff
@@ -94,6 +167,9 @@ void PeepholeOptimizeONNX(std::shared_ptr<Graph>& graph) {
   // TODO: make it easier not to do O(k) iterations over the graph, where
```
Comment (houseroad) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: Yes, we should design an API, so we can easily merge the iterations.

**Review by houseroad — APPROVED** [review]
Thanks, this is great!

BTW, what happens to our CI?

File: torch/csrc/jit/passes/onnx/peephole.cpp [hunk 3]
```diff
@@ -15,6 +15,24 @@ std::unordered_set<NodeKind> broadcasting = {
...
+bool isNopTranspose(const std::vector<int64_t> & perm) {
+  for (size_t i = 0; i < perm.size(); i++)
+    if (perm[i] != i)
+      return false;
+  return true;
+}
+
+// returns a vector `ret` such that transposing by `ret` is equivalent
+// to transposing by `t1` and then by `t2`
+std::vector<int64_t> composeTransposes(const std::vector<int64_t> & t1,
```
Comment (dzhulgakov) [inline torch/csrc/jit/passes/onnx/peephole.cpp]: double-check that the indices are valid and in range? (just in case)

**Comment by dzhulgakov** [comment]
Looks great! As for the tests - it seems that the ordering of ops in backward pass is not deterministic (https://travis-ci.org/pytorch/pytorch/jobs/299841429). It's not related to this diff, but @zdevito and @ezyang might be interested in figuring out a better way to unittest graphs.

**Comment by dzhulgakov** [comment]
Are we going to land it (I guess the only missing part is tests) or do you plan to put it in the separate transforms library?

**Comment by anderspapitto** [comment]
I would say no reason not to land it - and also some tests are here https://github.com/ezyang/onnx-pytorch/pull/40

**Pull request merged**